gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-common/src/main/java/io/gravitee/am/gateway/handler/common/vertx/web/handler/impl/internal/mfa/MFAEnrollStep.java

```diff
@@ -47,38 +54,82 @@ public MFAEnrollStep(Handler<RoutingContext> wrapper, RuleEngine ruleEngine, Fac
-        if (skipMFA) {
-            flow.doNext(routingContext);
+    private void required(RoutingContext routingContext, AuthenticationFlowChain flow, MfaFilterContext context) {
+        if (userHasFactor(context)) {
+            continueFlow(routingContext, flow);
         } else {
-            flow.exit(this);
+            enrollment(routingContext, flow);
         }
     }
 
-    private boolean noFactor(Client client) {
-        final Set<String> factors = client.getFactors();
-        return isNull(factors) || factors.isEmpty() || onlyRecoveryCodeFactor(factors);
+    private void conditional(RoutingContext routingContext, AuthenticationFlowChain flow, Client client, MfaFilterContext context) {
+        if (enrollConditionSatisfied(client, context)) {
+            stop(routingContext, flow);
+        } else if (userHasFactor(context)) {
+            continueFlow(routingContext, flow);
+        } else { //todo AM-1140 skip conditional not implemented
+            enrollment(routingContext, flow);
+        }
     }
 
-    private boolean onlyRecoveryCodeFactor(Set<String> factors) {
-        if (factors.size() == 1) {
-            final String factorId = factors.stream().findFirst().get();
-            final Factor factor = factorManager.getFactor(factorId);
-            return factor.getFactorType().equals(FactorType.RECOVERY_CODE);
+    private void optional(RoutingContext routingContext, AuthenticationFlowChain flow, MfaFilterContext context) {
+        if (context.isEnrollSkipped()) {
+            stop(routingContext, flow);
+        } else {
+            if (userHasFactor(context)) {
+                continueFlow(routingContext, flow);
+            } else {
+                enrollment(routingContext, flow);
+            }
         }
-        return false;
+    }
+
+    private void enrollment(RoutingContext routingContext, AuthenticationFlowChain flow) {
+        redirect(routingContext, flow, this);
+    }
+
+    private boolean enrollConditionSatisfied(Client client, MfaFilterContext context) {
+        return evaluateRule(getEnrollSettings(client).getEnrollmentRule(), context, ruleEngine);
+    }
+
+    private boolean isEnrollActive(Client client) {
+        return ofNullable(client.getMfaSettings()).map(MFASettings::getEnroll).map(EnrollSettings::isActive).orElse(false);
     }
 
     private boolean userHasFactor(MfaFilterContext context) {
-        return context.hasEndUserAlreadyEnrolled() || context.userHasMatchingFactors();
+        return context.isEndUserEnrolling() || context.userHasMatchingActivatedFactors();
     }
 
-    public boolean isMfaSkipped(MfaFilterContext context) {
-        // We need to check whether the AMFA rule is false since we don't know
-        final boolean mfaSkipped = context.isMfaSkipped();
-        final String mfaStepUpRule = context.getStepUpRule();
-        return isNullOrEmpty(mfaStepUpRule) && mfaSkipped;
+    public boolean canUserSkip(Client client, MfaFilterContext context) {
```

Comment:
**question**: this method is marked as Unused by IDEA... it is still useful ? It can be useful to keep it anyway (or create another one if this one is still useful) to keep in the RoutingContext (and maybe into the session) the fact that the user can't skip the enrollment in order to display or not the skip button on the enrollment page but also to check in the Enroll handler if the user was [allowed to skip](https://github.com/gravitee-io/gravitee-access-management/blob/master/gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/resources/endpoint/mfa/MFAEnrollEndpoint.java#L198C14-L198C30)

Reply:
yes, it will be used in the next story conditional enrollment skip